python/az/aro/azext_aro/custom.py

```diff
@@ -338,3 +301,92 @@ def get_network_resources(cli_ctx, subnets, vnet):
     resources.update(route_tables)
 
     return resources
+
+
+# service_principal_update manages cluster service principal update
+# 1. If called without parameters it should be best-effort
+# 2. If called with parameters it fails if something is not possible
+# 2.1. If client_id is provided, we expect expect secret to be provided too
+#      If only secret is provided - we are updating secret
+#      In any case we will try to verify and update rbac
+# 2.2. TODO: refresh-cluster-service-principal is provided
+def service_principal_update(cli_ctx, oc, client_id=None, client_secret=None):
+    rp_client_sp = None
+    client_sp = None
+    resources = set()
+
+    # if any of these are set - we expect users to have access to fix rbac so we fail
+    fail = client_id is not None or client_secret is not None
+
+    # update client_id without providing secret is not valid.
+    # this acts as dynamic validator
+    if client_id is not None:
+        if client_id != oc.service_principal_profile.client_id and client_secret is None:
+            raise InvalidArgumentValueError("Must specify --client-id with --client-secret.")
+
+    # if only secret is provided, we assume we re-use existing application
+    # it is users responsibility to vet it.
+    if client_id is None:
+        client_id = oc.service_principal_profile.client_id
+
+    if rp_mode_production():
+        rp_client_id = FP_CLIENT_ID
+    else:
+        rp_client_id = os.environ.get('AZURE_FP_CLIENT_ID', FP_CLIENT_ID)
+
+    try:
+        # Get cluster resources we need to assign network contributor on
+        resources = get_cluster_network_resources(cli_ctx, oc)
+    except (CloudError, HttpOperationError) as e:
+        raise logger.error(e.message) if fail else logger.info(e.message)
```

Comment:
this pattern is hard to read and also doesn't look right.  Something like:
```python
if fail:
    logger.error(e.message)
    raise

logger.info(e.message)
```

Reply:
I think contrary. I prefer single liner  for "best effort vs fail" . If you insist, I can change but I somehow think like my version better :)

Reply:
```
>>> import logging
>>> fail
>>> logging.error(error) if fail else logging.warning(error)
WARNING:root:error
>>> fail="set"
>>> logging.error(error) if fail else logging.warning(error)
ERROR:root:error
>>> fail=None
>>> logging.error(error) if fail else logging.warning(error)
WARNING:root:error
```

Reply:
Or Am I missing something?